Google Sheets Integration - Feature Branch

escherize opened the pull request:
## Overview

This PR establishes the foundational flow and backend endpoints for **Google Sheets Integration** Milestone 2. The focus is to get the core functionality working, and giving users a MVP to connect their Google Sheets Data as tables inside the app.

Showing the button in the first place is gated by a number of things. The `show-google-sheets-integration` setting, which requires being hosted on cloud, having an attached datawarehouse, having the etl connections feature, and having a mb-api-token. 

This way we ensure we can test and iterate on the feature safely.


# How it works
 ## Steps to connect a new google drive folder 
 
 - An admin clicks Add Data > Connect Google Sheets
 - FE sends :get "api/ee/gsheets/service-account" to get service-account email
 - If the FE cannot find the service-account email, it does not show the "connect to google sheets" (ctgs) button.
 - user clicks ctgs button -> FE shows instructions for how to connect a folder:
 - user shares their Google Drive folder w/ service-account email
 - user copies their Google Drive share link into MB form and hits submit
 - FE sends :post "api/ee/gsheets/folder" w/ folder url string as `request.body.url`
 - BE forwards request to :post "/api/v2/mb/connections" w/ body:
 `{:type "gdrive" :secret {:resources ["the-url"]}}`
 - on unexceptional status:
 - BE sets `gsheets.status` to `"loading"`
 - BE returns the gsheets shape: `{:status "loading" :folder_url "the-url" :folder-upload-time <epoch-time> :gdrive/conn-id <uuid>}}`
 - on exceptional status, BE returns a message like: "Unable to setup drive folder sync. Please check that the
 folder is shared with the proper service account email and sharing permissions." 
 
 ## Polling
 - FE polls :get "api/ee/gsheets/folder" until `body.status` == `complete`
 - BE forwards requests to :get "/api/v2/mb/connection/<gdrive-conn-id>", filtering for the google drive connection
 - If the connection doesn't exist: `{:error "google drive connection not found."}`
 - If the connection exists and is NOT active, return loading state in the response
 - If the connection exists and is active AND the latest sync of the attached datawarehouse is AFTER the sync time
 on Harbormaster's gdrive connection, return `body.status` == "complete".
 - FE sees status == "complete" 
 
 ### What does "complete" mean?
 When the status is "complete", the user can access their google sheets info inside Metabase.
 This means that both:
 1. the gdrive connection has been marked active
 2. the attached datawarehouse has completed syncing inside metabase
 To check this, we compare the last sync time of the attached datawarehouse with the last sync time of the gdrive. 
 
 ### What if the sync takes too long or never ends?
 Upon the successful creation of a gdrive connection, we start a timer that will error out if the sync does not
 complete within [[*folder-setup-timeout-seconds*]] seconds. If the sync does not complete within this window, we
 reset the gsheets status to `{:status "not-connected"}` and return an error for the FE to show. 
 
 ## Steps to disconnect a google drive folder
 - FE sends request to :delete "/folder"
 - MB loops over all gdrive connection ids and deletes them 
 
 ## Why do we need to sync the attached datawarehouse before considering it ready?
 We need to sync the attached datawarehouse to make sure that the data from the Google Drive folder is available to
 the user in Metabase. Once the gdrive connection's status is set to 'active' by HM, they will call MB's
 `api/notify/db/attached_datawarehouse` endpoint to trigger a sync. The data from the Google Drive folder is already
 availiable in the attached datawarehouse, and when MB finishes the sync (and puts an item into :model/TaskHistory
 saying so), the user can start using their Google Sheets data in Metabase. 
 
 ## Why do we check for multiple gdrive connections in the delete endpoint? We check for multiple gdrive connections
 in the delete endpoint to make sure that we are deleting _all_ gdrive connections. As of Milestone 2 there is
 supposed to be only one gdrive connection, but we observed multiple being created (distributed systems amirite?) and
 want the delete endpoint to basically reset us to a blank slate. If there are multiple gdrive connections, we will
 delete all of them. If there are errors during the gdrive conn deletion we ignore them .

## UI

. | .
-- | --
add data sidebar menu | ![Screen Shot 2025-03-03 at 5 18 08 PM](https://github.com/user-attachments/assets/6032958b-c060-4c1b-ac2d-22cf7b9ba541)
add data from db page | ![Screen Shot 2025-03-03 at 5 50 12 PM](https://github.com/user-attachments/assets/7ef0f658-7c45-458b-b500-d096ca6df7f7)
setup modal. | ![Screen Shot 2025-03-03 at 5 18 12 PM](https://github.com/user-attachments/assets/d5d664c8-126f-450c-b0f1-ce6813221cad)
sync status notification | ![Screen Shot 2025-03-03 at 5 19 08 PM](https://github.com/user-attachments/assets/2260746c-07f5-4fbc-84f0-c6020edc87e5)
success notification | ![Screen Shot 2025-03-03 at 5 45 44 PM](https://github.com/user-attachments/assets/f5af8dd3-7807-4c87-9c19-856d7c9cda9e)

iethree committed:
add frontend for gsheets sync

escherize committed:
add fall-through for ignored embedding requests

escherize committed:
Gsheets mb api (#51307)

* send flag to show "connect gsheets" button to FE

* clean up docstring + add etl-connections condition

* metabase api WIP

* align APIs

* include site-url in the HM temp-url api call

* cleanup

* update HM<->MB api to be working (minus temp-url)

* cleanup

* remove some notes

* indentation + linting

* add tests

* fix typo

* remove illustrative (when true ...)

escherize committed:
Merge branch 'master' into gsheets_integration

iethree committed:
update frontend to use changed API

escherize committed:
Merge branch 'master' into gsheets_integration

escherize committed:
Gsheets integration: more api alignment (#51492)

* add frontend for gsheets sync

* update frontend to use changed API

* wip

* cleanup

---------

Co-authored-by: Ryan Laurie <iethree@gmail.com>

escherize committed:
Merge branch 'master' into gsheets_integration

filipesilva committed:
fix: declare gsheet setting fn before using in getter

filipesilva committed:
fix: delay u.secret instantiation to prevent build race condition

See https://metaboat.slack.com/archives/CKZEMT1MJ/p1734605793293989

filipesilva committed:
build: push ee-extra to dev repo, not ee

Cloud wants ee-extra, not just ee

escherize committed:
move settings out of auth middleware ns

- update kondo modules
- add metabase.api.auth ns

escherize committed:
More cleaning up

- update docstring
- remove declares + use {set,get}-value-of-type{!,}
- remove nil state from gsheets setting

escherize committed:
rely on more hm stubs

escherize committed:
save folder_owner to gsheets

escherize committed:
gsheets and hm http client => enterprise

+ move tests too

escherize committed:
require the right namespace

escherize committed:
add migration for is_gsheet to metabase_table

escherize committed:
Gsheets service account be flow (#51997)

* update setup flow to use service-account

* align GET api/ee/gsheets/service-account response

* align the HM service account endpoint

* fix migration yaml

* update test for service accounts

* indentation + migration quibble

* improve error messages + kondo housekeeping

* remove unreachable if-let branch + fix serdes test

* add is_gsheet to mock and test data

* add gsheets setting to serialization test

* fix fe-type-check

* typechecker

escherize committed:
Merge remote-tracking branch 'origin' into gsheets_integration

escherize committed:
[kondo] defendpoint -> defendpoint2

escherize committed:
Merge branch 'master' into gsheets_integration

escherize committed:
Gsheets table management (#52145)

* include dwh ∈ GET /api/ee/upload-management/tables

* cljfmt

* simplify preliminary gsheet definition

* archive cards is a query param

escherize committed:
gsheets: update polling logic + better errors (#52162)

* update polling logic + better errors

* codify protocol for polling folder done-ness

* align on FE polling logic + the real sync complete

* [WIP] testing happy path

* status should be "active"

- remove a bunch of extra debugging

* remove outdated tests + cljfmt

* check admin + use the connection's id w/ delete

- cleanup

escherize committed:
fix :delete /folder + schema on :get /folder

iethree committed:
Gsheets frontend (#51946)

* non-oauth UI

* update api and add disconnect modal

* update illustration and add error displays

* add gsheets sync status

* better copy

* remove test code'

* fix bad api error

* updates to sync status

* Update sidebar entrypoint

filipesilva committed:
Merge branch 'master' into gsheets_integration

escherize committed:
gate things behind the right features

- only show gsheets button
- http requests to /api/ee/gsheets/*

escherize committed:
model/Table has a db_id column not database_id

escherize committed:
add and use simple events

escherize committed:
differentiate failure modes

escherize committed:
+ tests and error status codes

iethree committed:
make form submit on enter

escherize committed:
reset gsheet status when there is no gdrive conn

escherize committed:
add debug info

iethree committed:
FIXME FOR TESTING - disable feature checks

escherize committed:
fail on sync error + partially allow deletes

iethree committed:
add database page entrypoint

escherize committed:
FIXME FOR TESTING - set token-status-cache-ttl to 1 second (vs 12 hrs)

escherize committed:
only show _active_ dwh tables

escherize committed:
fix config.edn (howd it break tho?)

escherize committed:
handle multiple gdrive connections

- on delete, delete them all
- when looking at syncing, look at the most recent one

escherize committed:
add auto-cruft-tables to metabase_database.setting

- Here's the contract:

There should be an array of strings located at: db > settings >
"auto-cruft-tables". Each string will get fed into re-pattern. The table
names get lowercased before getting matched on. That list of regexes is
combined with our built in list of crufty regex patterns. We use re-find
to match them against the lower-cased table names. If there's a match we
will mark the table crufty (this is called a "Hidden Table" in the UI).
This way, if someone names their sheet `I'mUnique____dlt_loads` or
whatever they can at least fish it out of the hidden section.

Note: It will get set back to being crufty (aka what the ui calls
hidden) on the next sync... but this is configurable so we can fine-tune
the regexes to match the tables we want to hide more closely.

Examples:

If you put ["a"] in the config.yml: any table with A or a in it will be
marked hidden. So be sure to use anchors for the suffix or what-not. For
our thing, I think we want something like:

escherize committed:
add bunch of tests for auto-cruft

escherize committed:
mocked api tests

escherize committed:
add commment + mock hm responses

escherize committed:
Merge branch 'master' into gsheets_integration

escherize committed:
documentation

escherize committed:
Merge branch 'gsheets_integration' of github.com:metabase/metabase into gsheets_integration

escherize committed:
when multiple gdrive conns, sort by status, then created-at time

escherize committed:
cleanup

escherize committed:
Merge remote-tracking branch 'origin/master' into gsheets_integration

escherize committed:
joda time Instant/ofEpochSecond 0 -> t/instant 0

escherize committed:
if no gdrive conns are found, reset state

- add gdrive sorting tests

escherize committed:
indentation

iethree committed:
fix complete sync state and add error state

iethree committed:
update sync unit tests

iethree committed:
update sync unit tests

iethree committed:
add frontend analytics events

iethree committed:
don't show reconnect copy from DB page

escherize committed:
Add 10 minute timeout for polling connection

- return errors as json-encoded :error in the response body
- minor cleanup how we auto cruft tables (no change to behavior)
- add script for running the flow locally: from the repl

iethree committed:
better FE sync state and form validation

iethree committed:
refresh setting on error

iethree committed:
remove fake error state

escherize committed:
Delete connections on reset

- Change order of timeout checks in :get /folder handler
- add auto-cruft-columns setting

escherize committed:
Merge remote-tracking branch 'origin/master' into gsheets_integration

escherize committed:
update gsheets routes to use the new ns-handler

iethree committed:
loosen up gdrive link validation

iethree committed:
update regex

escherize committed:
add docstring to dynamic timeout var

escherize committed:
simplify api use via gdrive conn id

- move ->config into hm.client ns

escherize committed:
recover from suddenly-missing gdrive conn @ id

escherize committed:
cleanup hm.client

- indentation

fix [[make-request]] args

escherize committed:
fail to load invalid + present cruft patterns

escherize committed:
remove some extra comments + use hm-delete-conn

escherize committed:
fix update-table-metadata

- (no longer update nil -> nil)

escherize committed:
auto cruft columns

- pull the regex comparrison code into crufty ns
  - reuse it on 2
- add defsetting so having the setting on the database doesn't complain
- pass database down into field syncing + grab vis types

escherize committed:
indent

escherize committed:
fix a few tests

escherize committed:
fix gsheets tests

escherize committed:
add harbormaster client test

escherize committed:
kondo + cljfmt

escherize committed:
migrate to underscores for auto_cruft_*

- add tests for field + fix schema for table syncs

escherize committed:
use kebab-case for settings but accept snake_cased

escherize committed:
cljfmt + kebab case stragglers

escherize committed:
fix update-field-metadata-if-needed! signature

escherize committed:
fixing tests

escherize committed:
rename macro double -> run-twice

iethree committed:
remove test code

iethree committed:
prettier

crisptrutski committed:
Fix search index update job restart issue (#53229)

escherize committed:
cruft settings are db-local

escherize committed:
Merge branch 'gsheets_integration' of github.com:metabase/metabase into gsheets_integration

escherize committed:
more idiomatic crufty db-local settings

escherize committed:
Merge branch 'gsheets_integration' of github.com:metabase/metabase into gsheets_integration

escherize committed:
Merge remote-tracking branch 'origin/master' into gsheets_integration

escherize committed:
add auto cruft fields test

escherize committed:
bump gsheets sync timeout to 10 seconds

iethree committed:
handle going back to loading state

escherize committed:
kondo + add empty cruft tables test

escherize committed:
add get folder timeout test, update dox,

- remove regex folder check

escherize committed:
add cruft columns config.yml setup tests

escherize committed:
remove def + actually run the test

iethree committed:
test code for bruyan

escherize committed:
reset gsheets status on every error response

escherize committed:
Merge branch 'gsheets_integration' of github.com:metabase/metabase into gsheets_integration

iethree committed:
save dbId and actually show it

escherize committed:
bust cache on gsheets setting

escherize committed:
Merge branch 'gsheets_integration' of github.com:metabase/metabase into gsheets_integration

escherize committed:
Merge branch 'master' into gsheets_integration

escherize committed:
solve initial-sync-status-table-only-test failure

escherize committed:
Merge branch 'gsheets_integration' of github.com:metabase/metabase into gsheets_integration

rafpaf commented on enterprise/frontend/src/metabase-enterprise/google_sheets/GoogleSheetManagement.tsx:
```diff
@@ -0,0 +1,400 @@
+  const { status } = gSheetsSetting;
+
+  if (
+    !FIXME_TEST_BOOLEAN &&
```
^

rafpaf commented on enterprise/frontend/src/metabase-enterprise/google_sheets/GoogleSheetManagement.tsx:
```diff
@@ -0,0 +1,400 @@
+    useGetServiceAccountQuery();
+
+  if (
+    !FIXME_TEST_BOOLEAN &&
```
^

rafpaf commented on enterprise/frontend/src/metabase-enterprise/google_sheets/GoogleSheetManagement.tsx:
```diff
@@ -0,0 +1,400 @@
+  const gSheetsEnabled = useSetting("show-google-sheets-integration");
+
+  if (
+    !FIXME_TEST_BOOLEAN &&
```
^

rafpaf commented on enterprise/frontend/src/metabase-enterprise/google_sheets/GoogleSheetManagement.tsx:
```diff
@@ -0,0 +1,400 @@
+      <Text size="lg" fw="bold">
+        {
+          // eslint-disable-next-line no-literal-metabase-strings -- admin only string
+          t`Share the Google Drive folder that contains your Google Sheets with Metabase`
```
Maybe: `Share with Metabase the Google Drive folder that contains your Google Sheets`?

rafpaf commented on enterprise/frontend/src/metabase-enterprise/upload_management/index.ts:
```diff
@@ -1,8 +1,25 @@
+const FIXME_TEST_BOOLEAN = true;
+if (
+  (hasPremiumFeature("hosting") && hasPremiumFeature("attached_dwh")) ||
+  FIXME_TEST_BOOLEAN
```
^

rafpaf commented on frontend/src/metabase/nav/containers/MainNavbar/MainNavbarContainer/MainNavbarView.tsx:
```diff
@@ -172,7 +172,8 @@ export function MainNavbarView({
   const canUpload = canCurateRootCollection && canUploadToDatabase;
-  const showUploadCSVButton = hasAttachedDWHFeature && canUpload;
+  // const showUploadCSVButton = hasAttachedDWHFeature && canUpload;
+  const showUploadCSVButton = true; // FIXME for testing
```
^

rafpaf commented on frontend/src/metabase/status/components/StatusListing/StatusListing.tsx:
```diff
@@ -2,9 +2,11 @@ import { useBeforeUnload } from "react-use";
 import { hasActiveUploads } from "metabase/redux/uploads";
 import { getUserIsAdmin } from "metabase/selectors/user";
 import { useCheckActiveDownloadsBeforeUnload } from "metabase/status/hooks/use-check-active-downloads-before-unload";
+// FIXME: setup an enterprise plugin
```
^

rafpaf reviewed (commented):
Note that the FE code still has some FIXMEs and some (easily fixed) security issues pointed out by Github's automated tool

noahmoss commented on enterprise/backend/src/metabase_enterprise/gsheets.clj:
```diff
@@ -0,0 +1,350 @@
+  (hm.client/make-request :delete (str "/api/v2/mb/connections/" conn-id)))
+
+(defn- reset-gsheets-status []
+  (gsheets! not-connected)
+  (doseq [{:keys [id]} (hm-get-gdrive-conns)]
+    (hm-delete-conn id))
```
What happens here if these get out of sync? i.e. we reset the gsheets state in Metabase but HM fails to delete all the connections? Is that a recoverable state?

escherize replied:
That is the only way we will leak connections. It is still a possibility if hm fails to delete it, but I think this is good enough.

I am deleting all gdrive conns (though we use 1 and only 1) so that we don't leak connections.

If a user is disconnecting/reconnecting their drive folders then every time they disconnect (or hit an error), we try to clean up orphanted gdrive conns.

escherize commented on src/metabase/sync/sync_metadata/tables.clj:
```diff
@@ -115,13 +132,16 @@
-                                         :active false)]
-    ;; if the table already exists but is marked *inactive*, mark it as *active*
-    (t2/update! :model/Table existing-id (assoc (cruft-dependent-columns (:name table)) :active true))
+  (if-let [existing-id #p (t2/select-one-pk :model/Table
+                                            :db_id (u/the-id database)
+                                            :schema schema
+                                            :name table-name
+                                            :active false)]
+    (do   #p (t2/select-one :model/Table existing-id)
```
fixed, thank you.

escherize committed:
Update src/metabase/sync/sync_metadata/crufty.clj

Co-authored-by: Noah Moss <32746338+noahmoss@users.noreply.github.com>

escherize committed:
Update enterprise/backend/src/metabase_enterprise/advanced_config/file/databases.clj

Co-authored-by: Noah Moss <32746338+noahmoss@users.noreply.github.com>

escherize committed:
Update enterprise/backend/src/metabase_enterprise/advanced_config/file/databases.clj

Co-authored-by: Noah Moss <32746338+noahmoss@users.noreply.github.com>

escherize commented on enterprise/backend/src/metabase_enterprise/advanced_config/file/databases.clj:
```diff
@@ -25,15 +28,40 @@
+  (s/and
+   map?
+   (fn cruft-patterns-are-valid? [settings]
+     ;; _'s because this is validated first then normalized later
```
Thanks, I'll expand on it.

escherize committed:
Update enterprise/backend/src/metabase_enterprise/gsheets.clj

Co-authored-by: Noah Moss <32746338+noahmoss@users.noreply.github.com>

escherize commented on enterprise/backend/src/metabase_enterprise/gsheets.clj:
```diff
@@ -0,0 +1,350 @@
+            (filter is-gdrive? (or (some-> response :body) [])))
+      [])))
+
+(mu/defn- hm-delete-conn :- ::hm.client/http-reply
```
Done (not taking the commit because I renamed it inside the file).

escherize commented on enterprise/backend/src/metabase_enterprise/gsheets.clj:
```diff
@@ -0,0 +1,350 @@
+  :visibility :public
+  :type :json
+  :getter (fn [] (or
+                  ;; This NEEDS to be up to date between instances on a cluster, so:
+                  ;; we are going around the settings cache:
+                  (some-> (t2/select-one :model/Setting :key "gsheets") :value json/decode+kw)
```
We observed this while running as a cluster during an inner-loop trial.

When running on a cluster, if instance A gets the `:post /folder` request, and 5 seconds later instance B gets the `:get /folder` request, instance B will likely not have the (: I got bit by the settings cache for a couple hours. The ttl for refreshing the settings cache [is 1 minute](https://github.com/metabase/metabase/blob/gsheets_integration/src/metabase/models/setting/cache.clj#L132). 

I see that there's a binding I can set that will go around the cache, too.

We _do_ return gsheets in the system/properties response. It may be worth looking into exactly when we need to get the db-value and do it only there. I will check with @iethree about exactly when he needs the gsheets setting to be 100% reliable.

escherize committed:
respond to Noah's review comments

noahmoss commented on enterprise/backend/src/metabase_enterprise/gsheets.clj:
```diff
@@ -0,0 +1,351 @@
+  ;; TODO add api/+check-superuser +auth and +check-setting
+  (api.macros/ns-handler *ns*))
+
+(comment
```
want to remove this or are you including intentionally?

escherize replied:
fixed

noahmoss commented on enterprise/backend/src/metabase_enterprise/gsheets.clj:
```diff
@@ -0,0 +1,351 @@
+
+(def ^{:arglists '([request respond raise])} routes
+  "`/api/ee/gsheets` routes."
+  ;; TODO add api/+check-superuser +auth and +check-setting
```
Is this still needed?

escherize replied:
fixed

iethree committed:
update modal copy

noahmoss commented on src/metabase/sync/sync_metadata/tables.clj:
```diff
@@ -91,10 +98,20 @@
+                                database
+                                sync-stage :- [:enum ::reactivate ::create ::update]]
+  ;; if this is a crufty table, mark initial sync as complete since we'll be skipping the subsequent sync steps
+  (let [is-crufty? (crufty/name? table-name {:patterns        crufty-table-patterns
+                                             :pattern-strings (some-> database :settings :auto-cruft-tables)})]
```
It might be slightly cleaner for this API to just take a set regexes? And combine them all into one set before passing them in?

escherize replied:
good idea, done

iethree committed:
update error copy

noahmoss commented on src/metabase/sync/sync_metadata/tables.clj:
```diff
@@ -91,10 +98,20 @@
+                            ;; while updating its table metadata, so that it is "complete" during the sync.
+                            ;; See: [[metabase.sync.util-test/initial-sync-status-table-only-test]]
+                            (= sync-stage ::update) (:initial_sync_status table)
+                            ;; if a table is crufty, we mark it as complete to skip the subsequent sync steps
+                            is-crufty?              "complete"
```
does this actually work? if so I'm slightly surprised, I thought the sync status was just to power the UI

escherize replied:
Hmm, I believed this comment, but maybe it's no longer true (or never was). I dug around to see where the comment came from originally. [it is from this commit](https://github.com/metabase/metabase/blob/ffbb526e5deabbb668cc62cf363af2f4ecaac720/src/metabase/sync/sync_metadata/tables.clj#L99). I don't see anywhere that we are filtering for a table's initial sync status in the sync steps. 

So: I think it does get synced, but just marked crufty (hidden) which is fine for this since we want to allow admins to unhide tables that get marked crufty incidentally.

escherize replied:
I'll cleanup the comment

escherize committed:
kondo

iethree committed:
status click target update

escherize committed:
improve crufty/name? api

escherize committed:
Merge branch 'gsheets_integration' of github.com:metabase/metabase into gsheets_integration

escherize committed:
modularize gsheets

escherize committed:
remove temp gsheets

- token status checks happen once per 12h
- remove hm/conn and mb/sync-info from :get /folder response

escherize committed:
gdrive-conn is now unused

escherize committed:
Merge remote-tracking branch 'origin/master' into gsheets_integration

escherize committed:
remove comment

escherize committed:
Merge branch 'master' into gsheets_integration

escherize committed:
Merge branch 'gsheets_integration' of github.com:metabase/metabase into gsheets_integration

escherize committed:
fix modules

escherize committed:
add docstring

escherize committed:
add harbormaster module

escherize committed:
kondo

escherize committed:
fix the tests

escherize committed:
module linter

escherize committed:
fix typo

escherize committed:
Merge branch 'master' into gsheets_integration

escherize committed:
cljfmt

escherize committed:
setup api-key setting before reading it

escherize committed:
sort ns

escherize committed:
::snowplow/simple_event -> :snowplow/simple_event

iethree committed:
hard code error message

iethree committed:
remove test code

iethree committed:
update frontend tests

escherize committed:
update module linter config

escherize committed:
Merge branch 'master' into gsheets_integration

escherize committed:
add simple schema snowplow event schema

escherize committed:
fix schema name

escherize committed:
update some tests

escherize committed:
add a test

escherize committed:
better docstring

escherize committed:
better docstrings

iethree committed:
fix e2e tests

iethree committed:
fix fe token mock

escherize committed:
Merge remote-tracking branch 'origin/master' into gsheets_integration

escherize commented on src/metabase/sync/sync_metadata/crufty.clj:
```diff
@@ -0,0 +1,26 @@
+(defn ->regex
+  "Converts a pattern-string into a lower-cased regex."
+  [pattern]
+  (-> pattern u/lower-case-en re-pattern))
```
Thanks, I can see the confusion here. The intent is to make cruftyness regexes case insenstive so they work between table or column names that are upper and or lower case.

escherize committed:
respond to review comments

+ fix timeout 50000 -> 5000
+ make internal regex function private

filipesilva committed:
build: push ee-extra to dev repo, not ee

Cloud wants ee-extra, not just ee

iethree committed:
fix strings

escherize committed:
Merge branch 'master' into gsheets_integration

iethree committed:
remove resolved comment

escherize committed:
adds prometheus metrics to gsheets api

- :metabase-gsheets/sync-creation-begin
- :metabase-gsheets/sync-creation-ok
- :metabase-gsheets/sync-creation-error
  - failure-reason is one of:
  - "timeout" - gsheets.api/sync-complete? didn't return true after 10 minutes
  - "status_error" - HM said the sync connection had status == error
- :metabase-gsheets/connection-deleted

npfitz reviewed (approved):
Looks good to me!

escherize committed:
Merge remote-tracking branch 'origin/master' into gsheets_integration

escherize committed:
Merge branch 'gsheets_integration' of github.com:metabase/metabase into gsheets_integration

escherize committed:
Refines prometheus events

They are now:

- :metabase-gsheets/connection-creation-began
- :metabase-gsheets/connection-creation-ok
- :metabase-gsheets/connection-creation-error {:reason "status_error"}
- :metabase-gsheets/connection-creation-error {:reason "timeout"}
- :metabase-gsheets/connection-deleted

escherize committed:
comment

escherize committed:
allow resolution of snake_cased setting names

escherize committed:
Merge remote-tracking branch 'origin/master' into gsheets_integration

escherize commented on enterprise/backend/test/metabase_enterprise/api/gsheets_test.clj:
```diff
@@ -0,0 +1,211 @@
+      (is (= "Missing api-key."
+             (:message (mt/user-http-request :crowberto :get 500 "ee/gsheets/service-account")))))))
+
+(def happy-responses (read-string (slurp (io/resource "gsheets/mock_hm_responses.edn"))))
```
gsheets/mock_hm_responses.edn is an edn map from {:method _ :url "" :body _} to the response observed from HM.

I've added a few functions like +syncing that overwrite pieces of the map with other observed responses.

escherize commented on enterprise/backend/src/metabase_enterprise/api/routes.clj:
```diff
@@ -61,6 +64,9 @@
    "/autodescribe"               (premium-handler 'metabase-enterprise.llm.api :llm-autodescription)
    "/billing"                    metabase-enterprise.billing.api.routes/routes
    "/data-editing"               (premium-handler metabase-enterprise.data-editing.api/routes :table-data-editing)
+   "/gsheets"                    (-> gsheets.api/routes ;; gsheets requires both features.
+                                     (premium-handler :attached-dwh)
+                                     (premium-handler :etl-connections))
```
all gsheets routes are gated behind these 2 token features.

escherize commented on enterprise/backend/src/metabase_enterprise/gsheets/settings.clj:
```diff
@@ -0,0 +1,75 @@
+   [metabase.util.malli.schema :as ms]
+   [toucan2.core :as t2]))
+
+(mr/def :gsheets/gsheets
```
This isn't actually used, but it is good for docs

escherize replied:
I think the comment is good enough. I put a stub for "add malli schemas to defsetting?" to the march 20th guild sync

escherize commented on src/metabase/api/auth.clj:
```diff
@@ -0,0 +1,11 @@
+  (:require
+   [metabase.models.setting :refer [defsetting]]))
+
+(defsetting api-key
+  "When set, this API key is required for all API requests."
+  :encryption :when-encryption-key-set
+  :visibility :internal
+  :doc "Middleware that enforces validation of the client via the request header X-Metabase-Apikey.
+        If the header is available, then it’s validated against MB_API_KEY.
+        When it matches, the request continues; otherwise it’s blocked with a 403 Forbidden response.")
```
this got moved for modularity's sake

filipesilva committed:
Merge branch 'master' into gsheets_integration

escherize commented on enterprise/backend/src/metabase_enterprise/gsheets/api.clj:
```diff
@@ -0,0 +1,334 @@
+;;
+;; ## Steps to disconnect a google drive folder
+;; - FE sends request to :delete "/folder"
+;;   - MB loops over all gdrive connection ids and deletes them
```
You'd think there could only 1, but it's possible for there to be more in rare cases.

noahmoss commented on enterprise/backend/src/metabase_enterprise/harbormaster/client.clj:
```diff
@@ -0,0 +1,98 @@
+
+(mr/def :hm-client/http-reply [:tuple [:enum :ok :error] :map])
+
+(mu/defn make-request :- :hm-client/http-reply
```
wdyt of splitting this function up a bit? could have separate functions for `construct request`, `make-request`, and `handle-response`, or something like that? just an idea, not blocking

noahmoss commented on enterprise/backend/src/metabase_enterprise/harbormaster/client.clj:
```diff
@@ -0,0 +1,98 @@
+  `->config` either gets the store-api-url and api-key from settings or throws an exception when either are unset or
+  blank."
+  []
+  (let [store-api-url (setting/get-value-of-type :string :store-api-url)
```
any reason to use `get-value-of-type` and not just `(store-api-url)`? same with the other setting

escherize replied:
requiring store-api-url is a module violation, This only skirts the issue. I will require them directly and update the module linter config.

noahmoss commented on enterprise/backend/src/metabase_enterprise/harbormaster/client.clj:
```diff
@@ -0,0 +1,98 @@
+        _ (when (str/blank? store-api-url)
+            (log/error "Missing store-api-url. Cannot create hm client config.")
+            (throw (ex-info (tru "Missing store-api-url.") {:store-api-url store-api-url})))
+        api-key (setting/get-value-of-type :string :api-key)
```
also asked on Slack — but just want to check that we're intentionally overloading this `api-key` setting for this use case. I thought it's supposed to be a legacy env var specifically for the `/notify` endpoints

escherize replied:
It is that, but we also are using it to allow harbormaster to identify which metabase instance (or cluster, really) is setting up their ghsheets.

noahmoss commented on enterprise/backend/test/metabase_enterprise/harbormaster/client_test.clj:
```diff
@@ -0,0 +1,28 @@
+(deftest ->config-blank-api-key-test
+  (let [grv setting/get-raw-value]
+    ;; mt/with-temporary-setting-values is not used here because we want to test the behavior of the function when the
+    ;; api-key is blank or nil, and mt/with-temporary-setting-values will not allow us to set the api-key to blank
```
would `mt/with-temporary-raw-setting-values` work here?

escherize replied:
works for "", fails for nil. 🫠  I'll replace the "" case

noahmoss commented on enterprise/backend/test/metabase_enterprise/api/gsheets_test.clj:
```diff
@@ -0,0 +1,211 @@
+    (is (= :ok status))
+    (is (malli= [:map [:email [:string {:min 1}]]] (:body response)))))
+
+(def ^:private
+  gdrive-link
+  "nb: if you change this, change it in test_resources/gsheets/mock_hm_responses.edn"
+  "<gdrive-link>")
```
is this supposed to be an actual link?

noahmoss replied:
oh, I guess it's just the value in the mocked response? maybe calling it `expected-gdrive-link` or something like that would be clearer

noahmoss reviewed (changes requested):
Left a bunch of comments — mostly nits/style things but my biggest question is around the use of the `api-key` setting. I thought we would deprecate that eventually so I'm surprised that it's being used for a very different use case here. If this is an intentional decision I'll stamp 👍

escherize commented on enterprise/backend/src/metabase_enterprise/gsheets/api.clj:
```diff
@@ -0,0 +1,334 @@
+;; MB <-> HM APIs
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(mu/defn- hm-service-account-email :- [:or [:= false] :string]
```
nope, good catch

escherize commented on enterprise/backend/src/metabase_enterprise/gsheets/api.clj:
```diff
@@ -0,0 +1,334 @@
+(mu/defn- hm-service-account-email :- [:or [:= false] :string]
+  "Checks to see if Google service-account is setup in harbormaster, and returns the email."
+  []
+  (let [[_status {:keys [body] :as _response}] (hm.client/make-request :get "/api/v2/mb/connections-google/service-account")]
```
That's sensible, let's.

escherize commented on enterprise/backend/src/metabase_enterprise/gsheets/api.clj:
```diff
@@ -0,0 +1,334 @@
+
+(defn- handle-get-folder [attached-dwh]
+  (let [conn-id (or (:gdrive/conn-id (gsheets))
+                    (log/warn "CACHE MISS ON GSHEETS")
+                    (some-> (t2/select-one :model/Setting :key "gsheets")
+                            :value
+                            json/decode+kw
+                            :gdrive/conn-id))
```
👍

escherize commented on enterprise/backend/src/metabase_enterprise/gsheets/api.clj:
```diff
@@ -0,0 +1,334 @@
+                            :value
+                            json/decode+kw
+                            :gdrive/conn-id))
+        [sstatus {conn :body}] (try (hm-get-gdrive-conn conn-id)
```
it's intended. I didn't want to shadow `status` on line 247

escherize commented on enterprise/backend/src/metabase_enterprise/gsheets/api.clj:
```diff
@@ -0,0 +1,334 @@
+  ([message data]
+   (throw (ex-info message (merge data {;; the `:errors true` bit informs the exception middleware to return the message
+                                        ;; in the body, even if we want to pass a stauts code. Without `:errors true`,
+                                        ;; sending a stuats code will ellide the message from the response.
```
fixed

escherize commented on enterprise/backend/src/metabase_enterprise/gsheets/api.clj:
```diff
@@ -0,0 +1,334 @@
+;;
+;; ## Overview
+;; Google Sheets Integration allows users to connect their Google Drive folders to Metabase. We sync any Google Sheets
+;; inside of those folders, and make them avaliable to the user in their attached data warehouse. This integration is
```
:+1:

escherize commented on enterprise/backend/src/metabase_enterprise/gsheets/api.clj:
```diff
@@ -0,0 +1,334 @@
+  (api/+check-superuser
+   (api.macros/ns-handler *ns*)))
+
+(comment
```
done

noahmoss reviewed (approved):
Clarified `api-key` usage on Slack

escherize committed:
review responses

escherize committed:
add logging if any part of the batch-delete fails

escherize committed:
remove unused schema

escherize committed:
Revert "remove unused schema"

This schema is actually used! :tada:

This reverts commit f5dbd78051c84c34cbd9e1e8c771c78bc41e7a52.

iethree committed:
update loki icon reference image

escherize committed:
switch off flaky test setup

iethree committed:
mark date filter picker tests flaky (#54866)
